earth_enterprise/src/common/FileAccessorPluginLoader.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 The Open GEE Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "FileAccessorPluginLoader.h"
+#include "POSIXFileAccessor.h"
+#include "notify.h"
+
+void ScanPluginDirectory(const std::string &pluginDir, std::vector<std::string> &files) {
+    boost::filesystem::path p(pluginDir);
+    boost::system::error_code errorCode;
+    boost::filesystem::directory_iterator start(p, errorCode);
+    if (errorCode.value() != 0){
+        notify(NFY_WARN, "Unable to scan file accessor plugin directory, %s. %s.", pluginDir.c_str(), errorCode.message().c_str());
+        return;
+    }
+
+    boost::filesystem::directory_iterator end;
+    std::transform(start, end, std::back_inserter(files), 
+        [pluginDir](const boost::filesystem::directory_entry& entry) {
+            return pluginDir + entry.path().leaf().string(); });
+}
+
+FileAccessorPluginLoader& FileAccessorPluginLoader::Get() {
+    static FileAccessorPluginLoader loader;
```

Comment:
Is there a possibility of multithreading here? We could end up creating multiple loaders if so.

Reply:
I think it's safe to do this in C++11. If I'm reading correctly, the standard guarantees that only one thread can execute the declaration at a time. (See section 6.7, paragraph 4).
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf
```
If control enters the declaration concurrently while the variable is being initialized, the concurrent execution shall wait for
completion of the initialization
```
But maybe I'm reading that wrong.

Reply:
Looks like you're right. We should be safe as long as the `FileAccessorPluginLoader` is thread-safe, which it is.